### PR TITLE
Better parser error recovery when encountering a open generic name.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/67277

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler  this is ready for review.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -5701,131 +5700,145 @@ private ScanTypeArgumentListKind ScanTypeArgumentList(NameOptions options)
         }
 
         private ScanTypeFlags ScanPossibleTypeArgumentList(
-            ref SyntaxToken lastTokenOfList, out bool isDefinitelyTypeArgumentList)
+            out SyntaxToken greaterThanToken,
+            out bool isDefinitelyTypeArgumentList)
         {
             isDefinitelyTypeArgumentList = false;
+            Debug.Assert(this.CurrentToken.Kind == SyntaxKind.LessThanToken);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:5707]: view with whitepace off.

this was always called from a caller that already checked we were on a `<`.  So no need to check/indent all this code.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -5701,131 +5700,145 @@ private ScanTypeArgumentListKind ScanTypeArgumentList(NameOptions options)
...
+                Debug.Assert(greaterThanToken.Kind == SyntaxKind.GreaterThanToken);
 
-                    // Type arguments cannot contain attributes, so if this is an open square, we early out and assume it is not a type argument
-                    if (this.CurrentToken.Kind == SyntaxKind.OpenBracketToken)
-                    {
-                        lastTokenOfList = null;
-                        return ScanTypeFlags.NotType;
-                    }
+                return ScanTypeFlags.GenericTypeOrMethod;
+            }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:5724]: this is hte new logic.

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson @jcouv ptal :)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -5701,131 +5700,145 @@ private ScanTypeArgumentListKind ScanTypeArgumentList(NameOptions options)
...
-                        // It would be nice if we saw a tuple to state that we definitely had a 
-                        // type argument list.  However, there are cases where this would not be
-                        // true.  For example:
+                if (this.CurrentToken.Kind == SyntaxKind.GreaterThanToken)
+                {
+                    greaterThanToken = EatToken();
+                    return result;
+                }
+
+                switch (this.ScanType(out _))
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:5745]: this out-param never contributed to the greaterThanToken.  it was always overwritten afterwards.  changed to `out _` to be actively less confusing.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler  ptal.

**Comment by CyrusNajmabadi** [comment]
@jcouv ptal.

**Comment by CyrusNajmabadi** [comment]
@333fred ptal :)

**Review by jjonescz — APPROVED** [review]

**Pull request merged**